document-store/src/main/java/org/hypertrace/core/documentstore/Collection.java

```diff
@@ -30,8 +30,11 @@ public interface Collection {
    *
    * @param key Unique key of the document in the collection.
    * @param document Document to be upserted.
-   * @return True if this operation resulted in update of an existing document. False, otherwise.
+   * @return True if this operation resulted in an update of an existing document or creation of a
+   *     new document.
+   * @deprecated Use {@link #createOrReplace} for a consistent behavior.
    */
+  @Deprecated(forRemoval = true)
```

Comment:
What is the difference between `upsert` and `createOrReplace`?
Are we facing some issue with the existing api? if, yes, what?

Reply:
Yes. The issue is mentioned in the documentation comments of the `upsert` API.
For example, upserting 
```json
{
    "foo2": "bar2"
}
```
into an existing document
```json
{
    "foo1": "bar1"
}
```
results in 
```json
{
    "foo2": "bar2"
}
```
correctly in Postgres, but, results in
```json
{
    "foo1": "bar1",
    "foo2": "bar2"
}
```
in Mongo implementation.

Reply:
- Was `updateSubDoc` capable to solve this problem? 
- I guess `upsert` was meant for full document update. Is the requirement to add some additional stuff in the existing document?

Reply:
> Was updateSubDoc capable to solve this problem?

Nopes. There also, I need to explicitly unset the other top-level fields that are undesired.

> I guess upsert was meant for full document update.

Yes. It was supposed to completely replace the existing document. But, that was not happening in the case of MongoDB.

> Is the requirement to add some additional stuff in the existing document?

Nope. The new API targets at fixing the bug in the Mongo implementation of upsert.

Reply:
As discuss offline, we need a way to move to usage of newer APIs. So, one of the way to define collectionV2.